File: sdk/go/loom/retry.go
package loom

import (
	"context"
	"errors"
	"math/rand"
	"time"
)

func (c *WebhookClient) doWithRetry(ctx context.Context, attemptFunc func(context.Context) (*TriggerResult, error)) (*TriggerResult, error) {
	var lastErr error
	baseDelay := 500 * time.Millisecond
	maxDelay := 5 * time.Second

	r := rand.New(rand.NewSource(time.Now().UnixNano()))

	for attempt := 0; attempt < c.maxRetries; attempt++ {
		res, err := attemptFunc(ctx)
		if err == nil {
			return res, nil
		}
		
		lastErr = err

		// Check context cancellation before delaying
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}

		var delay time.Duration

		var rateLimitErr *RateLimitedError
		if errors.As(err, &rateLimitErr) {
			delay = rateLimitErr.RetryAfter
		} else if isRetryableError(err) {
			// Exponential backoff
			backoff := time.Duration(float64(baseDelay) * float64(int(1)<<attempt))
			if backoff > maxDelay {
				backoff = maxDelay
			}
			jitter := time.Duration(r.Float64() * float64(time.Second))
			delay = backoff + jitter
		} else {
			// Non-retryable error (e.g. 400, 401, 404)
			return nil, err
		}

		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(delay):
			// continue to next attempt
		}
	}

	return nil, lastErr
}

func isRetryableError(err error) bool {
	var internalErr *InternalServerError
	if errors.As(err, &internalErr) {
		return true
	}
	
	var isErr *InvalidSignatureError
	if errors.As(err, &isErr) {
		return false
	}
	var wnErr *WebhookNotFoundError
	if errors.As(err, &wnErr) {
		return false
	}
	var ipErr *InvalidPayloadError
	if errors.As(err, &ipErr) {
		return false
	}
	
	// Network errors
	var apiErr *LoomAPIError
	if errors.As(err, &apiErr) {
		return false
	}
	
	return true
}

File: sdk/python/loom/retry.py
import time
import random
from typing import Callable, TypeVar
import requests
from requests.exceptions import RequestException
from .errors import (
    LoomAPIError,
    InternalServerError,
    RateLimitedError
)

T = TypeVar('T')

def do_with_retry(
    attempt_func: Callable[[], T],
    max_retries: int = 5,
    base_delay: float = 0.5,
    max_delay: float = 5.0
) -> T:
    last_err = None
    
    for attempt in range(max_retries):
        try:
            return attempt_func()
        except RateLimitedError as e:
            last_err = e
            delay = e.retry_after
        except (InternalServerError, RequestException) as e:
            last_err = e
            backoff = min(base_delay * (2 ** attempt), max_delay)
            jitter = random.uniform(0, 1.0)
            delay = backoff + jitter
        except LoomAPIError as e:
            # All other API errors (e.g., 400, 401, 404) are non-retryable
            raise e
            
        time.sleep(delay)
        
    raise last_err
